File: network/src/main/kotlin/org/bitlap/network/processor/GetColumnsProcessor.kt
package org.bitlap.network.processor

import com.alipay.sofa.jraft.rpc.RpcContext
import com.alipay.sofa.jraft.rpc.RpcRequestClosure
import com.google.protobuf.Message
import org.bitlap.network.core.NetworkService
import org.bitlap.network.core.SessionHandle
import org.bitlap.network.proto.driver.BGetColumns.BGetColumnsReq
import org.bitlap.network.proto.driver.BGetColumns.BGetColumnsResp
import java.util.concurrent.Executor

/**
 * GetColumns
 *
 * @author 梦境迷离
 * @since 2021/6/5
 * @version 1.0
 */
class GetColumnsProcessor(
    private val networkService: NetworkService,
    executor: Executor? = null,
) : BitlapRpcProcessor<BGetColumnsReq>(executor, BGetColumnsResp.getDefaultInstance()) {

    override fun processRequest(request: BGetColumnsReq, done: RpcRequestClosure): Message {
        val result =
            networkService.getColumns(
                SessionHandle(request.sessionHandle),
                request.tableName,
                request.schemaName,
                request.columnName
            )
        return BGetColumnsResp.newBuilder()
            .setStatus(success()).setOperationHandle(result.toBOperationHandle()).build()
    }

    override fun processError(rpcCtx: RpcContext, exception: Exception): Message {
        return BGetColumnsResp.newBuilder().setStatus(error(exception)).build()
    }

    override fun interest(): String = BGetColumnsReq::class.java.name
}

File: network/src/main/scala/org/bitlap/net/processor/GetColumnsProcessor.scala
package org.bitlap.net.processor

import com.alipay.sofa.jraft.rpc.{ RpcContext, RpcRequestClosure }
import com.google.protobuf.Message
import org.bitlap.net.{ NetworkService, handles }
import org.bitlap.network.proto.driver.BGetColumns.{ BGetColumnsReq, BGetColumnsResp }

import java.util.concurrent.Executor

/**
 *
 * @author 梦境迷离
 * @since 2021/11/21
 * @version 1.0
 */
class GetColumnsProcessor(private val networkService: NetworkService, executor: Executor = null)
  extends BitlapRpcProcessor[BGetColumnsReq](executor, BGetColumnsResp.getDefaultInstance)
    with ProcessorHelper {

  override def processRequest(request: BGetColumnsReq, done: RpcRequestClosure): Message = {
    val result =
      networkService.getColumns(
        new handles.SessionHandle(request.getSessionHandle),
        request.getTableName,
        request.getSchemaName,
        request.getColumnName
      )
    BGetColumnsResp.newBuilder()
      .setStatus(success()).setOperationHandle(result.toBOperationHandle()).build()
  }

  override def processError(rpcCtx: RpcContext, exception: Exception): Message = BGetColumnsResp.newBuilder().setStatus(error(exception)).build()

  override def interest(): String = classOf[BGetColumnsReq].getName

}
